File: 07-数据存储/02-NoSQL/ScyllaDB分片感知与shard-per-core/python/sharding.py
"""ScyllaDB token -> shard 映射（对拍自 scylladb/scylladb dht/token.cc）。

源码口径（master 分支，实读）：

    inline unsigned
    zero_based_shard_of(uint64_t token, unsigned shards, unsigned sharding_ignore_msb_bits) {
        token <<= sharding_ignore_msb_bits;               // uint64 左移，高位丢弃
        return (uint128_t(token) * shards) >> 64;
    }

    unsigned
    shard_of(unsigned shard_count, unsigned msb, const token& t) {
        switch (t._kind) {
            case before_all_keys: return token::shard_of_minimum_token();   // 0
            case after_all_keys:  return shard_count - 1;
            case key:             return zero_based_shard_of(unbias(t), shard_count, msb);
        }
    }

    constexpr uint64_t unbias() const { return uint64_t(_data) + uint64_t(INT64_MIN); }
"""

MASK64 = (1 << 64) - 1
INT64_MIN = -(1 << 63)


class Token:
    """dht::token：三态（before_all_keys / key / after_all_keys），数据部分是 int64。"""

    BEFORE = "before_all_keys"
    KEY = "key"
    AFTER = "after_all_keys"

    def __init__(self, kind, data=0):
        self.kind = kind
        self.data = data

    @staticmethod
    def key(value):
        return Token(Token.KEY, value)

    @staticmethod
    def bias(n):
        """源码：token(kind::key, n - uint64(INT64_MIN)) —— unbias 的逆。"""
        v = (n - (1 << 63)) & MASK64  # 先按 uint64 取模
        if v >= (1 << 63):            # 再按 int64 解释（有符号回绕）
            v -= 1 << 64
        return Token(Token.KEY, v)

    @classmethod
    def minimum(cls):
        return cls(cls.BEFORE, 0)

    @classmethod
    def maximum(cls):
        return cls(cls.AFTER, 0)

    def is_minimum(self):
        return self.kind == Token.BEFORE

    def is_maximum(self):
        return self.kind == Token.AFTER

    def raw(self):
        # 源码：maximum 时 raw() 返回 int64 max，其他情况返回 _data
        return (1 << 63) - 1 if self.is_maximum() else self.data

    def unbias(self):
        # uint64(_data) + uint64(INT64_MIN)  —— 把 int64 环挪到 [0, 2^64)
        return ((self.data + INT64_MIN) & MASK64) if not self.is_maximum() else 0

    def __repr__(self):
        return "Token(%s,%d)" % (self.kind, self.data) if self.kind == Token.KEY else "Token(%s)" % self.kind


def zero_based_shard_of(token_u64, shards, msb):
    """主函数：把 token 当 [0,1) 的小数，shard = floor(0.token * shards)。"""
    shifted = (token_u64 << msb) & MASK64
    return ((shifted * shards) >> 64) & MASK64


def shard_of(shard_count, msb, t):
    """dht::shard_of —— 三态分派。"""
    if t.is_minimum():
        return 0  # token::shard_of_minimum_token()，源码注释 "hardcoded for now"
    if t.is_maximum():
        return shard_count - 1
    return zero_based_shard_of(t.unbias(), shard_count, msb)


def init_zero_based_shard_start(shards, msb):
    """zero_based_shard_of 的逆：ret[s] = 属于 s 的最小 token。"""
    if shards == 1:
        return [0]
    ret = []
    for s in range(shards):
        token = (s << 64) // shards
        token >>= msb
        # 除法向下取整可能落到前一个 shard，逐格 +1 修正到边界
        while zero_based_shard_of(token, shards, msb) != s:
            token += 1
        ret.append(token)
    return ret


def sharding_algorithm_name():
    return "biased-token-round-robin"


class StaticSharder:
    """dht::static_sharder（vnode 表）：对同一个配置，token -> shard 恒定不变。"""

    def __init__(self, shard_count, msb=0):
        self.shard_count = shard_count
        self.msb = msb
        self.shard_start = init_zero_based_shard_start(shard_count, msb)

    def shard_of(self, t):
        return shard_of(self.shard_count, self.msb, t)

    def shard_for_reads(self, t):
        return self.shard_of(t)

    def shard_for_writes(self, t, migrating=False):
        """tablet 迁移期间本地写要同时写给旧、新两个 shard（shard_replica_set 容量 2）。"""
        s = self.shard_of(t)
        if migrating:
            other = (s + 1) % self.shard_count
            return sorted({s, other}) if other != s else [s]
        return [s]

    def token_for_next_shard(self, t, shard, spans=1):
        """返回 > t 的、属于 shard 的第一个分片边界 token；溢出返回 maximum。"""
        if t.is_maximum() or shard >= self.shard_count:
            return Token.maximum()
        n = t.unbias() if not t.is_minimum() else 0
        s = zero_based_shard_of(n, self.shard_count, self.msb)
        # 源码语义：目标 shard 必须严格在当前 shard 之后（不回绕），否则视为溢出
        if self.msb == 0:
            n = self.shard_start[shard]
            if spans > 1 or shard <= s:
                return Token.maximum()
        else:
            left = (n >> (64 - self.msb)) + spans - (1 if shard > s else 0)
            if left >= (1 << self.msb):
                return Token.maximum()
            n = (left << (64 - self.msb)) | self.shard_start[shard]
        return Token.bias(n)

    def next_shard(self, t):
        shard = self.shard_for_reads(t)
        nxt = shard + 1 if shard + 1 != self.shard_count else 0
        tok = self.token_for_next_shard(t, nxt)
        if tok.is_maximum():
            return None
        return (nxt, tok)


class FixedShardPartitioner:
    """dht::fixed_shard_partitioner：token 编码 [shard:16][hash:48]，给 Raft 元数据表用。"""

    shard_bits = 16
    shard_shift = 64 - 16
    max_shard = 32767  # int16 max，源码写死
    hash_mask = (1 << shard_shift) - 1

    @classmethod
    def token_for_shard(cls, shard, hash_bits):
        value = (shard << cls.shard_shift) | (hash_bits & cls.hash_mask)
        return Token.key(value)

    @classmethod
    def shard_of(cls, t):
        return (t.raw() >> cls.shard_shift) & MASK64

    @classmethod
    def shard_of_clamped(cls, t, shard_count):
        """sharder 里额外 clamp：min(shard, shard_count - 1)。"""
        return min(cls.shard_of(t), shard_count - 1)

File: 07-数据存储/02-NoSQL/ScyllaDB分片感知与shard-per-core/go/sharding.go
// Package main 转写自 scylladb/scylladb 的 dht/token.cc 与 dht/fixed_shard.cc。
// C++ 的 uint128 乘法在 Go 里用 math/big 落地，避免溢出。
package main

import (
	"fmt"
	"math"
	"math/big"
)

const mask64 = uint64(math.MaxUint64)

// Token 对应 dht::token：三态，数据部分是 int64。
type Token struct {
	Kind string // "before_all_keys" / "key" / "after_all_keys"
	Data int64
}

const (
	kindBefore = "before_all_keys"
	kindKey    = "key"
	kindAfter  = "after_all_keys"
)

func keyToken(v int64) Token { return Token{Kind: kindKey, Data: v} }
func minToken() Token        { return Token{Kind: kindBefore} }
func maxToken() Token        { return Token{Kind: kindAfter} }

func (t Token) isMinimum() bool { return t.Kind == kindBefore }
func (t Token) isMaximum() bool { return t.Kind == kindAfter }

func (t Token) raw() int64 {
	if t.isMaximum() {
		return math.MaxInt64
	}
	return t.Data
}

// unbias 对应 C++：uint64(_data) + uint64(INT64_MIN)，Go 里无符号加法天然回绕。
func (t Token) unbias() uint64 {
	if t.isMaximum() {
		return 0
	}
	return uint64(t.Data) + uint64(math.MinInt64)
}

// bias 是 unbias 的逆：n - uint64(INT64_MIN)，再按 int64 解释。
func bias(n uint64) Token {
	v := int64(n - uint64(math.MinInt64))
	return keyToken(v)
}

// zeroBasedShardOf 对应源码主函数：floor((0.token) * shards)。
func zeroBasedShardOf(token uint64, shards, msb uint) uint64 {
	shifted := token << msb // uint64 左移，高位丢弃
	prod := new(big.Int).SetUint64(shifted)
	prod.Mul(prod, big.NewInt(int64(shards)))
	prod.Rsh(prod, 64)
	return prod.Uint64()
}

func shardOf(shardCount, msb uint, t Token) uint64 {
	switch {
	case t.isMinimum():
		return 0 // token::shard_of_minimum_token()
	case t.isMaximum():
		return uint64(shardCount - 1)
	default:
		return zeroBasedShardOf(t.unbias(), shardCount, msb)
	}
}

// initZeroBasedShardStart 是 zeroBasedShardOf 的逆：ret[s] 为属于 s 的最小 token。
func initZeroBasedShardStart(shards, msb uint) []uint64 {
	if shards == 1 {
		return []uint64{0}
	}
	ret := make([]uint64, shards)
	for s := uint(0); s < shards; s++ {
		num := new(big.Int).Lsh(big.NewInt(1), 64)
		num.Mul(num, big.NewInt(int64(s)))
		num.Div(num, big.NewInt(int64(shards)))
		token := num.Uint64() >> msb
		for zeroBasedShardOf(token, shards, msb) != uint64(s) {
			token++
		}
		ret[s] = token
	}
	return ret
}

// StaticSharder 对应 dht::static_sharder。
type StaticSharder struct {
	ShardCount uint
	MSB        uint
	ShardStart []uint64
}

func NewStaticSharder(shardCount, msb uint) *StaticSharder {
	return &StaticSharder{
		ShardCount: shardCount,
		MSB:        msb,
		ShardStart: initZeroBasedShardStart(shardCount, msb),
	}
}

func (s *StaticSharder) ShardOf(t Token) uint64 {
	return shardOf(s.ShardCount, s.MSB, t)
}

func (s *StaticSharder) ShardForReads(t Token) uint64 { return s.ShardOf(t) }

// ShardForWrites tablet 迁移期要同时写旧、新两个 shard。
func (s *StaticSharder) ShardForWrites(t Token, migrating bool) []uint64 {
	sh := s.ShardOf(t)
	if !migrating {
		return []uint64{sh}
	}
	other := (sh + 1) % uint64(s.ShardCount)
	if other == sh {
		return []uint64{sh}
	}
	if other < sh {
		return []uint64{other, sh}
	}
	return []uint64{sh, other}
}

// TokenForNextShard 目标 shard 必须严格在当前之后，否则视为溢出返回 maximum。
func (s *StaticSharder) TokenForNextShard(t Token, shard, spans uint) Token {
	if t.isMaximum() || shard >= s.ShardCount {
		return maxToken()
	}
	var n uint64
	if !t.isMinimum() {
		n = t.unbias()
	}
	cur := zeroBasedShardOf(n, s.ShardCount, s.MSB)
	if s.MSB == 0 {
		n = s.ShardStart[shard]
		if spans > 1 || shard <= uint(cur) {
			return maxToken()
		}
	} else {
		left := (n >> (64 - s.MSB)) + uint64(spans)
		if uint64(shard) > cur {
			left--
		}
		if left >= uint64(1)<<s.MSB {
			return maxToken()
		}
		n = (left << (64 - s.MSB)) | s.ShardStart[shard]
	}
	return bias(n)
}

func (s *StaticSharder) NextShard(t Token) (uint64, Token, bool) {
	sh := s.ShardForReads(t)
	next := sh + 1
	if next == uint64(s.ShardCount) {
		next = 0
	}
	tok := s.TokenForNextShard(t, uint(next), 1)
	if tok.isMaximum() {
		return 0, Token{}, false
	}
	return next, tok, true
}

// FixedShardPartitioner token 编码 [shard:16][hash:48]。
type FixedShardPartitioner struct{}

const (
	fixedShardBits  = 16
	fixedShardShift = 64 - fixedShardBits
	fixedMaxShard   = math.MaxInt16
	fixedHashMask   = uint64(1<<fixedShardShift) - 1
)

func (FixedShardPartitioner) TokenForShard(shard uint16, hashBits uint64) Token {
	value := int64(uint64(shard)<<fixedShardShift | (hashBits & fixedHashMask))
	return keyToken(value)
}

func (FixedShardPartitioner) ShardOf(t Token) uint64 {
	return uint64(t.raw()) >> fixedShardShift
}

func (f FixedShardPartitioner) ShardOfClamped(t Token, shardCount uint) uint64 {
	sh := f.ShardOf(t)
	if sh > uint64(shardCount-1) {
		return uint64(shardCount - 1)
	}
	return sh
}

func (t Token) String() string {
	if t.Kind != kindKey {
		return fmt.Sprintf("Token(%s)", t.Kind)
	}
	return fmt.Sprintf("Token(key,%d)", t.Data)
}
